lib/core/src/display/scene/transform.rs

```diff
@@ -0,0 +1,84 @@
+use nalgebra::{Matrix4, Quaternion, UnitQuaternion, Vector3};
+
+// Note [Quaternion YXZ Euler angles]
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+// To comply with Threejs impl, we generate a Quaternion applying rotation in
+// the order: pitch -> roll -> yaw, instead of roll -> pitch -> yaw based on
+// https://github.com/mrdoob/three.js/blob/master/src/math/Quaternion.js#L199
+fn from_euler_angles_pry(roll: f32, pitch: f32, yaw: f32) -> UnitQuaternion<f32> {
+    let (s1, c1): (f32, f32) = (roll * 0.5 as f32).sin_cos();
+    let (s2, c2): (f32, f32) = (pitch * 0.5 as f32).sin_cos();
+    let (s3, c3): (f32, f32) = (yaw * 0.5 as f32).sin_cos();
+
+    UnitQuaternion::from_quaternion(Quaternion::new(
+        c1 * c2 * c3 - s1 * s2 * s3,
+        s1 * c2 * c3 + c1 * s2 * s3,
+        c1 * s2 * c3 - s1 * c2 * s3,
+        c1 * c2 * s3 + s1 * s2 * c3,
+    ))
+}
+
+pub struct Transform {
+    pub position:   Vector3<f32>,
+    pub quaternion: UnitQuaternion<f32>,
+    pub scale:      Vector3<f32>,
+}
+
+impl Transform {
+    pub fn identity() -> Self {
+        Self {
+            position:   Vector3::new(0.0, 0.0, 0.0),
+            quaternion: UnitQuaternion::identity(),
+            scale:      Vector3::new(1.0, 1.0, 1.0),
+        }
+    }
+
+    pub fn set_position(&mut self, x: f32, y: f32, z: f32) {
+        self.position = Vector3::new(x, y, z);
+    }
+
+    pub fn set_scale(&mut self, x: f32, y: f32, z: f32) {
+        self.scale = Vector3::new(x, y, z);
+    }
+
+    pub fn set_rotation(&mut self, roll: f32, pitch: f32, yaw: f32) {
+        self.quaternion = from_euler_angles_pry(roll, pitch, yaw);
+    }
+
+    // Note [Transform to Matrix4 composition]
+    // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    // based on https://github.com/mrdoob/three.js/blob/master/src/math/Matrix4.js#L732
```

Comment:
This should be a doc comment

Reply:
I left this as a source comment and provided a detailed explanation on how it works